src/openforms/prefill/service.py

```diff
@@ -0,0 +1,160 @@
+            continue
+        if not prefill.get("plugin"):
+            continue
+        if not prefill.get("attribute"):
+            continue
+
+        default_value = component.get("defaultValue")
+        # 1693: we need to normalize values according to the format expected by the
+        # component. For example, (some) prefill plugins return postal codes without
+        # space between the digits and the letters.
+        prefill_value = normalize_value_for_component(component, prefill_value)
+
+        if prefill_value != default_value and default_value is not None:
+            logger.info(
+                "Overwriting non-null default value for component %r",
+                component,
+            )
+        component["defaultValue"] = prefill_value
+
+
+@elasticapm.capture_span(span_type="app.prefill")
+def prefill_variables(submission: Submission, register: Registry | None = None) -> None:
+    """Update the submission variables state with the fetched attribute values.
+
+    For each submission value variable that need to be prefilled, the according plugin will
+    be used to fetch the value. If ``register`` is not specified, the default registry instance
+    will be used.
+    """
+    from openforms.formio.service import normalize_value_for_component
+
+    from .registry import register as default_register
+
+    register = register or default_register
+
+    state = submission.load_submission_value_variables_state()
+    variables_to_prefill = state.get_prefill_variables()
+
+    component_variables: list[SubmissionValueVariable] = []
+    user_defined_variables: list[SubmissionValueVariable] = []
+    prefill_data: defaultdict[str, JSONEncodable] = defaultdict(dict)
+
+    for variable in variables_to_prefill:
+        if (
+            variable.form_variable.source == FormVariableSources.component
+            and variable.form_variable.prefill_attribute
+        ):
+            component_variables.append(variable)
+        elif (
+            variable.form_variable.source == FormVariableSources.user_defined
+            and variable.form_variable.prefill_options
+        ):
+            user_defined_variables.append(variable)
+
+    if component_variables and (
+        component_results := fetch_prefill_values_for_component(
+            submission, register, component_variables
+        )
+    ):
+        prefill_data.update(**component_results)
+    if user_defined_variables and (
+        user_defined_results := fetch_prefill_values_for_user_defined(
+            submission, register, user_defined_variables
+        )
+    ):
+        prefill_data.update(**user_defined_results)
+
+    total_config_wrapper = submission.total_configuration_wrapper
+    for variable_key, prefill_value in prefill_data.items():
+        component = total_config_wrapper[variable_key]
+        normalized_prefill_value = normalize_value_for_component(
+            component, prefill_value
+        )
```

Comment:
since `prefill_data` includes user defined variables, do we have guards here for missing `component` for a given key?